pkg/scheduler/plugins/gpushare/pod_handler.go

```diff
@@ -0,0 +1,128 @@
+		pod = t
+	case cache.DeletedFinalStateUnknown:
+		var ok bool
+		pod, ok = t.Obj.(*corev1.Pod)
+		if !ok {
+			return
+		}
+	default:
+		return
+	}
+	podRequest, _ := resource.PodRequestsAndLimits(pod)
+	if isGPUPod(podRequest) {
+		n.unAssume(nil, pod, pod.Spec.NodeName)
+	}
+}
+
+// add is used to update deviceUsed when there is a new pod created
+func (n *nodeDeviceCache) assume(allocations []*apiext.DeviceAllocation, pod *corev1.Pod, nodeName string) {
+	n.nodeDeviceInfos[nodeName].lock.Lock()
+	defer n.nodeDeviceInfos[nodeName].lock.Unlock()
+
+	if len(allocations) != 0 {
+		for _, allocation := range allocations {
+			(*n.nodeDeviceInfos[nodeName].deviceUsed[schedulingv1alpha1.GPU])[int(allocation.Minor)] = quotav1.Add(
+				(*n.nodeDeviceInfos[nodeName].deviceUsed[schedulingv1alpha1.GPU])[int(allocation.Minor)],
+				allocation.Resources)
+		}
+	} else {
+		devicesAllocation, err := apiext.GetDevicesAllocation(pod)
+		if err != nil {
+			klog.Errorf("failed to get %v device allocation, err: %v", pod.Name, err)
+			return
+		}
+		gpuAllocations := (*devicesAllocation)[schedulingv1alpha1.GPU]
+		for _, allocation := range gpuAllocations {
+			(*n.nodeDeviceInfos[nodeName].deviceUsed[schedulingv1alpha1.GPU])[int(allocation.Minor)] = quotav1.Add(
+				(*n.nodeDeviceInfos[nodeName].deviceUsed[schedulingv1alpha1.GPU])[int(allocation.Minor)],
+				allocation.Resources)
+
+		}
+	}
+
+	for minor, usedResources := range *n.nodeDeviceInfos[nodeName].deviceUsed[schedulingv1alpha1.GPU] {
+		(*n.nodeDeviceInfos[nodeName].deviceFree[schedulingv1alpha1.GPU])[minor] = quotav1.Subtract(
+			(*n.nodeDeviceInfos[nodeName].deviceTotal[schedulingv1alpha1.GPU])[minor],
+			usedResources)
+	}
+}
+
+// unAssume is used to update deviceUsed when there is pod deleted
+func (n *nodeDeviceCache) unAssume(allocations []*apiext.DeviceAllocation, pod *corev1.Pod, nodeName string) {
+	n.nodeDeviceInfos[nodeName].lock.Lock()
+	defer n.nodeDeviceInfos[nodeName].lock.Unlock()
+
+	if len(allocations) != 0 {
+		for _, allocation := range allocations {
+			(*n.nodeDeviceInfos[nodeName].deviceUsed[schedulingv1alpha1.GPU])[int(allocation.Minor)] = quotav1.SubtractWithNonNegativeResult(
+				(*n.nodeDeviceInfos[nodeName].deviceUsed[schedulingv1alpha1.GPU])[int(allocation.Minor)],
+				allocation.Resources)
+		}
+	} else {
+		devicesAllocation, err := apiext.GetDevicesAllocation(pod)
+		if err != nil {
+			klog.Errorf("failed to get %v device allocation, err: %v", pod.Name, err)
+			return
+		}
+		gpuAllocations := (*devicesAllocation)[schedulingv1alpha1.GPU]
+		for _, allocation := range gpuAllocations {
+			(*n.nodeDeviceInfos[nodeName].deviceUsed[schedulingv1alpha1.GPU])[int(allocation.Minor)] = quotav1.SubtractWithNonNegativeResult(
+				(*n.nodeDeviceInfos[nodeName].deviceUsed[schedulingv1alpha1.GPU])[int(allocation.Minor)],
+				allocation.Resources)
+		}
+	}
+
+	for minor, usedResources := range *n.nodeDeviceInfos[nodeName].deviceUsed[schedulingv1alpha1.GPU] {
```

Comment:
The "reset" free operation is not safe. bacause you should pay attention that total\used\free's dimension of minor may not same, for example if card has hardware-error,  the total may(or not) decrease some minors. if some minor decreases but still has used, the free should be negtive.

Reply:
This must be a really rare case. I use `quotav1.SubtractWithNonNegativeResult` to make sure we don't return negative values to prevent negative resource usage.